File: 2022/PythonSolutions/Day07.py
import re
from collections import defaultdict

with open("../Inputs/InputDay07.txt") as file:
    dirs, path = defaultdict(int), []
    commands = re.split(r'\n\$ ', '\n' + file.read())
    for cmd in commands:
        if cmd.startswith('cd'):
            path = path[:-1] if cmd == 'cd ..' else path + [cmd.removeprefix('cd ')]
        else:
            size = sum(int(x.split()[0]) for x in cmd.split('\n')[1:] if not x.startswith('dir'))
            for i in range(len(path)):
                dirs['/'.join(path[:i+1])] += size
    print(sum(x for x in dirs.values() if x <= 100000))
    print(min([x for x in dirs.values() if x >= dirs['/'] - 40000000]))

File: 2022/rust_solutions/src/day07.rs
use itertools::Itertools;
use regex::Regex;
use std::collections::HashMap;
use std::fs;

pub fn solve() {
    let data = fs::read_to_string("../Inputs/InputDay07.txt").expect("Unable to read file");
    let mut dirs: HashMap<String, usize> = HashMap::new();
    let mut path = vec![];
    let re = Regex::new(r"\n\$ ").unwrap();
    let data = format!("\n{}", data);
    let commands = re.split(data.as_str());
    for cmd in commands {
        if cmd == "cd .." {
            path.pop();
        } else if cmd.starts_with("cd") {
            path.push(cmd.strip_prefix("cd ").unwrap());
        } else {
            let size: usize = cmd
                .split('\n')
                .skip(1)
                .filter_map(|x| x.split(' ').next().unwrap().parse::<usize>().ok())
                .sum();
            for i in 0..path.len() {
                *dirs.entry(path.iter().take(i + 1).join("/")).or_default() += size;
            }
        }
    }
    println!(
        "{}",
        dirs.values().filter(|&x| *x <= 100000usize).sum::<usize>()
    );
    println!(
        "{}",
        dirs.values()
            .filter(|&x| *x >= dirs["/"] - 40000000usize)
            .min()
            .unwrap()
    );
}
